### ISSUE TITLE
Is the library thread-safe?

### ISSUE THREAD

**Issue by sivabudh** [fnc12/sqlite_orm#163]
Eg. multiple readers / writers are trying to access the same storage at the same time. Does the library arbitrate the accesses or I need to mutex it.

**Comment by sivabudh** [comment]
From this article,

https://www.sqlite.org/threadsafe.html

It says that the default mode is `serialized`. I'm running the application on Windows installed via `vcpkg`. So, I take it that SQLite is already in `serialized` mode?

**Comment by sivabudh** [comment]
The way I structured my code is that I made the class methods `static` so that any threads can access the database. Is that the best practice way?

Here's an example. 

```
// SomeDb.h
class SomeDb
{
public:
   struct Record
   {
         int id;
         std::string some_data;
         boo resolved;
   };
   static Record search_by(int id_);
};
```

and here's the implementation file
```
// SomeDb.cpp
auto make_sqlite_storage()
{
   using namespace sqlite_orm;
   auto storage = make_storage("somedb.sqlite",  make_table("Record", 
                                       make_column(/* the usual */) )  
                             );
    storage.sync_schema();
   return storage;
}

/******* !! Important !! using a single object holding connection to Sqlite *********/
auto g_Storage = make_sqlite_storage();

Record SomeDb::search_by(int id_)
{
    using namespace sqlite_orm;
    auto record = g_Storage.get_all<Record>(where(c(&Record::id) == id_));
    return record[0]; // assumes record always found for simplicity 
}
```

**Comment by fnc12** [comment]
1) `static` has nothing common with initial issue about thread safety cause you can both access your storage via static functions and via non-static member functions by passed reference to a storage from different threads. IMHO `static` should me omitted if it can be omitted cause it makes architecture more clean and mockable
2) if you want to access storage from different threads you should call `open_forever` right after making a storage cause storage can create different connections as a result of data races inside a storage
3) the only thing related to thread safety is `busy_timeout` member function in `storage_t` class. There is no inner mutex in storage and there is nothing about handling inner sqlite multithreading. So I'd advice you using your own mutex before accessing a database

But if you're interested we can investigate multithreading in sqlite. First of all I can add [threadsafe](https://www.sqlite.org/c3ref/threadsafe.html) function. You'll be able to check whether or not your storage is threadsafe.

**Comment by sivabudh** [comment]
Thanks. I think to have the threadsafe function to check storage is threadsafe is very helpful.

**Comment by sivabudh** [comment]
And further, so when you designed this library, did you intend for different threads to share the same storage reference or for each thread to hold its own storage?

**Comment by sivabudh** [comment]
I never knew about `open_forever`. May I conclude that if an application was multithreaded, that is a SQLite database file will be accessed by different threads simultaneously, the best practice is to _always_ need to call `open_forever`?

**Comment by sivabudh** [comment]
You may also be interested in this StackOverflow post:

https://stackoverflow.com/questions/10325683/can-i-read-and-write-to-a-sqlite-database-concurrently-from-multiple-connections

**Comment by sivabudh** [comment]
How would I enable WAL with sqlite_orm? 

https://blog.xojo.com/2015/12/21/speed-up-sqlite-with-write-ahead-logging/

**Comment by fnc12** [comment]
`open_forever` guarantees that any access to the database through a storage will be performed with one and only one connection

**Comment by fnc12** [comment]
Also I can ass `PRAGMA journal_mode` easily with `storage.pragma.journal_mode = ...` interface for WAL

**Comment by sivabudh** [comment]
OK, RE: `open_forever`, then instead of needing to use my own mutex, I could just call `open_forever` and all accesses will be serialized by sqlite_orm.

In other words, `open_forever` serializes all accesses for me. 

So do you recommend me using `open_forever` if I'm running into `std::system_error` due to `SQLITE_BUSY` or should I use WAL?

**Comment by sivabudh** [comment]
By the way,

I will be upgrading to [1.2](https://github.com/fnc12/sqlite_orm/releases/tag/v1.2) and trying out the new getters/setters feature. Will feedback if I run into any problems. =)

**Comment by fnc12** [comment]
you still will receive `SQLITE_BUSY` sometimes after `open_forever` cause you can write something while something other is being written by other thread. `open_forever` is required to just eliminate data races inside storage class itself, not sqlite database. Without calling `open_forever` one can create and set two different connection to `storage_t::current_transaction` member. This will lead to undefined behavior. Also you can increase busy timeout right now.
Also I'll add `threadsafe` member function to `storage_t` class and push it to a new branch. Also I'll add pragma `journal_mode` in that branch

**Comment by sivabudh** [comment]
Thanks, after considerations, I will put mutex around all accesses to SQLite.

FWIW, I will use a writer / reader mutex for better performance. See: http://doc.qt.io/qt-5/qreadwritelock.html

**Comment by sivabudh** [comment]
And thanks for adding `threadsafe` and `journal_mode`, I will be testing them via `vcpkg` when you make the release. =)

**Comment by fnc12** [comment]
I would be better if you test `threadsafe` and `journal_mode` before release right from their special branch or `dev` branch cause I need a feedback before merging a feature. And these two features are not enough for a new release

**Comment by sivabudh** [comment]
OK. I will test for you. Just let me know when it's pushed to Dev. Thank you.

**Comment by fnc12** [comment]
sorry I got sick and didn't finish these functions yet. I'll make it in a week

**Comment by fnc12** [comment]
@sivabudh hi. I've finally finished a branch with `threadsafe` and `journal_mode`. You can inspect it in https://github.com/fnc12/sqlite_orm/pull/165 . Also feel free to review it and/or comment it. Also you can pull a branch named `threadsafe-and-journal-mode` to test journal mode and threadsafe. BTW threadsafe is available already now by calling `sqlite3_threadsafe();`. This function takes no arguments, lol. I've created an interface for it: `int sqlite_orm::threadsafe()`.
Journal mode can be accessed with pragma member in storage:`storage.pragma.journal_mode(journal_mode::OFF);`

**Comment by fnc12** [comment]
@sivabudh I'm gonna merge #165 and close the issue

**Comment by fnc12** [comment]
merged #165
